### ISSUE TITLE
[Adv. Medical] Heal hitpoints doesn't work

### ISSUE THREAD

**Issue by Magirot** [acemod/ACE3#2498]
ACE3 Version: 3.3.1.2

**Mods:**
* `@CBA_A3`
* `@ace`

**Placed ACE3 Modules:**
* `Medical Settings (ACE) [Medical Level: Advanced, Medics setting: Advanced]`
* `Advanced Medical Settings (ACE) [Heal hitpoints: Yes]`

**Description:**
With the heal hitpoints parameter, bandaging wounds should heal the damage taken as well. This doesn't seem to work in the newest version.
Here's a video if it helps: https://youtu.be/5Xi2AGv3uCs

**Steps to reproduce:**
1. Add the aforementioned modules (or use an userconfig with the parameters).
2. Get shot and bandage yourself.
3. Still bloodied. :(

**Comment by Pergor** [comment]
Additionally you still can't run after bandaging up some leg hits with ace_medical_healHitPointAfterAdvBandage == 1.

If I can recall it right, it's the third time this was supposed to be fixed.

**Comment by Magirot** [comment]
Do you mean it didn't work in previous versions either? At least for us the damage got healed like it's supposed to in 3.2.1.

**Comment by Pergor** [comment]
No, the leg damage remained and you could naught but walk after bandaging up the third leg wound, even in 3.2.1.

**Comment by Glowbal** [comment]
Interesting. When I last tried this everything worked fine. I guess we need to take another look.

**Comment by Pergor** [comment]
These are the settings I currently use:

class ace_medical_level {
    value = 2;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_medicSetting {
    value = 2;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_enableFor {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_enableOverdosing {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_bleedingCoefficient {
    value = 0.3;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_painCoefficient {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_enableAirway {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_enableFractures {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_enableAdvancedWounds {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_enableVehicleCrashes {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_enableScreams {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_playerDamageThreshold {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_AIDamageThreshold {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_enableUnconsciousnessAI {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_remoteControlledAI {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_preventInstaDeath {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_enableRevive {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_maxReviveTime {
    value = 2700;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_amountOfReviveLives {
    value = -1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_allowDeadBodyMovement {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_allowLitterCreation {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_litterCleanUpDelay {
    value = 300;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_medicSetting_PAK {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_medicSetting_SurgicalKit {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_consumeItem_PAK {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_consumeItem_SurgicalKit {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_useLocation_PAK {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_useLocation_SurgicalKit {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_useCondition_PAK {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_useCondition_SurgicalKit {
    value = 0;
    typeName = "SCALAR";
    force = 1;
};
class ace_medical_keepLocalSettingsSynced {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_healHitPointAfterAdvBandage {
    value = 1;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_painIsOnlySuppressed {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_allowUnconsciousAnimationOnTreatment {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_moveUnitsFromGroupOnUnconscious {
    value = 0;
    typeName = "BOOL";
    force = 1;
};
class ace_medical_menu_allow {
    value = 1;
    typeName = "SCALAR";
    force = 1;
};

**Comment by jokoho48** [comment]
@Pergor please use pastbin or other text pasting pages thanks

**Comment by DerekSauer** [comment]
Looking at the code in "fnc_treatmentAdvanced_bandageLocal.sqf" starting at line 97, ACE is checking to see if all wounds on a particular part have been treated and then setting that part to undamaged on Arma's end of things with the setHitPointDamage command. 

ACE is removing damage from HitLeftArm, HitRightArm, HitLeftLeg and HitRightLeg but these parts do nothing in Arma's damage model. I think they're just bone locations for the animation system. Arma's damage system is much simpler and combines arms and legs into one 'Hit Point' so it only tracks damage to "HitArms" and "HitLegs", not each limb individually. 

You can test this yourself in game with the debug console. Apply damage to a limb with setHitPointDamage while you have ACE enabled (E.g.: player setHitPointDamage ["HitLeftLeg", 1.0];) and your player will continue to run around without issue and show no blood. Do the same with "HitLegs" and you will begin to walk around slowly and your pants will be bloody. Heal it with player setHitPointDamage ["HitLegs", 0.0]; and now your player will run around as normal and the blood will be gone.

I believe that ACE needs to check to see if all wounds to both legs or both arms has been treated and then use the correct setHitPointDamage (HitArms or HitLegs) to remove blood and arma engine injuries.

**Comment by DerekSauer** [comment]
I was thinking something like the following as an unsexy brute force fix:

```
// If all wounds have been bandaged, we will reset all damage to 0, so the unit is not showing any blood on the model anymore.
if (GVAR(healHitPointAfterAdvBandage) && {{(_x select 2) == _part && {((_x select 4) * (_x select 3)) > 0}}count _openWounds == 0}) then {
    _hitSelections = ["head", "body", "hand_l", "hand_r", "leg_l", "leg_r"];
    _hitPoints = ["HitHead", "HitBody", "HitLeftArm", "HitRightArm", "HitLeftLeg", "HitRightLeg"];
    _point = _hitPoints select (_hitSelections find _selectionName);
    _target setHitPointDamage [_point, 0];

    // Fix for "Arma-Injury" and bloody uniform
    // If both arms have been fully bandaged set Arma's aggregate arms to undamaged
    if ((_target getHitPointDamage "HitLeftArm" == 0) && (_target getHitPointDamage "HitRightArm" == 0)) then
    {
        _target setHitPointDamage ["HitArms", 0];
    };

    // If both legs have been fully bandaged set Arma's aggregate legs to undamaged
    if ((_target getHitPointDamage "HitLeftLeg" == 0) && (_target getHitPointDamage "HitRightLeg" == 0)) then
    {
        _target setHitPointDamage ["HitLegs", 0];
    };
};
```

**Comment by DerekSauer** [comment]
After testing this some more in game it looks like ACE never actually applies any damage to HitRightArm, HitRightLeg, etc... so having them healed by the code in healHitPointAfterAdvBandage wasn't doing anything in the first place, aside from those HitPoints not having any actual function. To that effect the code I posted above will always fully heal the Arma aggregate limbs when bandaged since HitRightArm, etc... always have no damage so getHitPointDamage will always return 0.

**Comment by DerekSauer** [comment]
Also have to correct myself. The aggregate for arms is not HitArms it is HitHands. I've corrected the snippet above even though its no longer useful.

**Comment by Glowbal** [comment]
Fixed in #2646 by @DerekSauer :+1:
